File: Reference/tests/Brontide.Reference.Interchange.Tests/Portable/PortableTestHarness.cs
using System.Collections.Immutable;
using System.Text.Json;
using System.Threading.Channels;
using Brontide.Reference.Experimental.Binding.Portable;

namespace Brontide.Reference.Interchange.Tests.Portable;

/// <summary>
/// Shared scaffolding for the portable-binding vectors: both realizations, the neutral artifacts,
/// and the value-description reader that turns a golden encoding's description into a shaped value.
/// </summary>
internal static class PortableTestHarness
{
    public static string NeutralRoot { get; } = Path.Combine(AppContext.BaseDirectory, "binding", "portable");

    public static JsonDocument ReadNeutral(params string[] relative)
    {
        var path = Path.Combine(new[] { NeutralRoot }.Concat(relative).ToArray());
        Assert.That(File.Exists(path), Is.True, $"The neutral artifact '{path}' was not copied to the test output.");
        return JsonDocument.Parse(File.ReadAllText(path));
    }

    /// <summary>Every vector id declared by the neutral layer, across all vector files.</summary>
    public static ImmutableArray<string> NeutralVectorIds()
    {
        var ids = ImmutableArray.CreateBuilder<string>();
        foreach (var file in Directory.GetFiles(Path.Combine(NeutralRoot, "vectors"), "*.json"))
        {
            using var document = JsonDocument.Parse(File.ReadAllText(file));
            if (!document.RootElement.TryGetProperty("vectors", out var vectors))
            {
                continue;
            }

            foreach (var vector in vectors.EnumerateArray())
            {
                ids.Add(vector.GetProperty("id").GetString()!);
            }
        }

        return ids.ToImmutable();
    }

    public static PortableProviderEndpoint CoolingEndpoint(
        PortableRealization realization,
        IPortableOperationHandler? handler = null,
        PortableContractDocument? contract = null) =>
        new(
            contract ?? CoolingPortableFixture.Contract,
            handler ?? new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry()),
            realization);

    /// <summary>The fixed direct-call realization: no wire, no encoding, no copy.</summary>
    public static ValueTask<PortableBindingHost> DirectHostAsync(
        PortableContractDocument? contract = null,
        IPortableOperationHandler? handler = null)
    {
        var document = contract ?? CoolingPortableFixture.Contract;
        var endpoint = new PortableProviderEndpoint(
            document,
            handler ?? new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry()),
            PortableRealization.FixedDirectCall);
        return PortableBindingHost.EstablishAsync(document, new PortableDirectConversation(endpoint), "reference-direct");
    }

    /// <summary>
    /// The negotiated process realization over a local duplex seam: real length-delimited framing
    /// and a real deterministic-CBOR encode and decode on both sides.
    /// </summary>
    public static async Task<(PortableBindingHost Host, PortableLocalSeam Seam)> ProcessHostAsync(
        PortableContractDocument? contract = null,
        IPortableOperationHandler? handler = null)
    {
        var document = contract ?? CoolingPortableFixture.Contract;
        var seam = PortableLocalSeam.Create(document.Limits);
        var endpoint = new PortableProviderEndpoint(
            document,
            handler ?? new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry()),
            PortableRealization.NegotiatedProcess);
        seam.StartProvider(endpoint, document.Limits);
        var host = await PortableBindingHost.EstablishAsync(
            document,
            new PortableProcessConversation(seam.HostDuplex, document.Limits),
            "reference-process");
        return (host, seam);
    }

    public static PortableConstraint Permitted() =>
        PortableConstraint.AllOf(
            PortableConstraint.Atom(PortableTruth.Satisfied),
            PortableConstraint.Atom(PortableTruth.Satisfied));

    public static PortableCopiedBlobResource Blob(string name = "profile", string content = "cooling-profile") =>
        PortableCopiedBlobResource.Create(name, System.Text.Encoding.UTF8.GetBytes(content));

    public static string Hex(ReadOnlySpan<byte> bytes) => Convert.ToHexStringLower(bytes);

    public static byte[] FromHex(string hex) => Convert.FromHexString(hex);

    /// <summary>
    /// Reads a golden encoding's explicit value description into a shaped value. The description
    /// carries kind discriminators; the wire form does not, which is the point of the comparison.
    /// </summary>
    public static PortableValue ReadDescribedValue(JsonElement element)
    {
        var kind = element.GetProperty("kind").GetString();
        switch (kind)
        {
            case "unit":
                return PortableUnitValue.Instance;
            case "text":
                return new PortableTextValue(element.GetProperty("value").GetString()!);
            case "boolean":
                return new PortableBooleanValue(element.GetProperty("value").GetBoolean());
            case "integer":
                return new PortableIntegerValue(element.GetProperty("value").GetInt64());
            case "decimal":
                return new PortableDecimalValue(
                    element.GetProperty("exponent").GetInt32(),
                    element.GetProperty("mantissa").GetInt64());
            case "bytes":
                return new PortableBytesValue(FromHex(element.GetProperty("hex").GetString()!));
            case "sequence":
                return new PortableSequenceValue(
                    [.. element.GetProperty("items").EnumerateArray().Select(ReadDescribedValue)]);
            case "choice":
                return new PortableChoiceValue(
                    element.GetProperty("case").GetString()!,
                    ReadDescribedValue(element.GetProperty("value")));
            case "record":
            {
                var record = PortableRecordValue.Empty;
                foreach (var field in element.GetProperty("fields").EnumerateObject())
                {
                    record = record.WithField(field.Name, ReadDescribedValue(field.Value));
                }

                foreach (var fragment in element.GetProperty("fragments").EnumerateObject())
                {
                    var reference = PortableFragmentReference.ParseText(fragment.Name);
                    record = record.WithFragment(
                        reference,
                        [.. fragment.Value.EnumerateObject()
                            .Select(field => (field.Name, ReadDescribedValue(field.Value)))]);
                }

                return record;
            }
            default:
                throw new InvalidOperationException($"The golden value description uses unknown kind '{kind}'.");
        }
    }

    public static PortableShapeReference ReadShapeReference(JsonElement element) =>
        PortableShapeReference.Parse(
            element.GetProperty("name").GetString()!,
            element.GetProperty("version").GetInt32());
}

/// <summary>Small edits over decoded control items, so an adversarial frame can be built exactly.</summary>
internal static class CborTestExtensions
{
    public static CborMap With(this CborMap map, string key, CborItem value) =>
        new([.. map.Entries.Where(entry => entry.Key != key).Append(new CborMapEntry(key, value))]);

    public static CborArray Array(this CborMap map, string key) =>
        (CborArray)map.Entries.Single(entry => entry.Key == key).Value;

    public static CborArray Replace(this CborArray array, int index, CborItem value) =>
        new(array.Items.SetItem(index, value));
}

/// <summary>
/// A duplex seam between a host and a provider endpoint inside one test process.
/// </summary>
/// <remarks>
/// It is a real pipe pair with real framing, so the process realization's encoding, bounds, and
/// copy accounting are exercised. It is not process isolation: the cross-process evidence lives in
/// the gated tests that launch the provider executable.
/// </remarks>
internal sealed class PortableLocalSeam : IAsyncDisposable
{
    private readonly SeamStream _hostToProvider;
    private readonly SeamStream _providerToHost;
    private readonly CancellationTokenSource _cancellation = new();
    private Task? _provider;

    private PortableLocalSeam(SeamStream hostToProvider, SeamStream providerToHost, PortableLimits limits)
    {
        _hostToProvider = hostToProvider;
        _providerToHost = providerToHost;
        HostDuplex = new PortableStreamDuplex(providerToHost, hostToProvider, limits, ownsStreams: false);
        ProviderDuplex = new PortableStreamDuplex(hostToProvider, providerToHost, limits, ownsStreams: false);
    }

    public IPortableDuplex HostDuplex { get; }

    public IPortableDuplex ProviderDuplex { get; }

    public static PortableLocalSeam Create(PortableLimits limits) =>
        new(new SeamStream(), new SeamStream(), limits);

    public void StartProvider(PortableProviderEndpoint endpoint, PortableLimits limits) =>
        _provider = Task.Run(
            () => PortableProviderProcessLoop.RunAsync(ProviderDuplex, endpoint, limits, _cancellation.Token),
            CancellationToken.None);

    /// <summary>Runs a deliberately misbehaving peer instead of the conforming endpoint.</summary>
    public void StartPeer(Func<IPortableDuplex, CancellationToken, Task> peer) =>
        _provider = Task.Run(() => peer(ProviderDuplex, _cancellation.Token), CancellationToken.None);

    /// <summary>
    /// Ends the provider's outbound direction, which the host observes as a clean end between
    /// frames rather than as an interrupted one.
    /// </summary>
    public void CloseProviderOutput() => _providerToHost.CompleteWriting();

    public async ValueTask DisposeAsync()
    {
        await _cancellation.CancelAsync();
        _hostToProvider.CompleteWriting();
        _providerToHost.CompleteWriting();

        try
        {
            if (_provider is not null)
            {
                await _provider.WaitAsync(TimeSpan.FromSeconds(5));
            }
        }
        catch (Exception)
        {
            // Teardown: a peer failing as its seam disappears is the expected outcome, and the test
            // has already asserted whatever it came to assert.
        }

        _cancellation.Dispose();
    }

    /// <summary>
    /// One direction of the seam: a byte stream whose reader blocks until bytes arrive, observes
    /// cancellation, and sees a clean end once the writer completes.
    /// </summary>
    /// <remarks>
    /// An anonymous pipe would be closer to the real transport, but its reads neither observe
    /// cancellation nor separate "writer closed" from "handle disposed" on Windows. Those are
    /// exactly the two distinctions these vectors turn on, so the seam models them explicitly.
    /// </remarks>
    private sealed class SeamStream : Stream
    {
        private readonly Channel<byte[]> _segments = Channel.CreateUnbounded<byte[]>();
        private byte[] _current = [];
        private int _offset;

        public override bool CanRead => true;
        public override bool CanSeek => false;
        public override bool CanWrite => true;
        public override long Length => throw new NotSupportedException();

        public override long Position
        {
            get => throw new NotSupportedException();
            set => throw new NotSupportedException();
        }

        public void CompleteWriting() => _segments.Writer.TryComplete();

        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
        {
            while (_offset == _current.Length)
            {
                if (!await _segments.Reader.WaitToReadAsync(cancellationToken))
                {
                    return 0;
                }

                if (_segments.Reader.TryRead(out var segment))
                {
                    _current = segment;
                    _offset = 0;
                }
            }

            var count = Math.Min(buffer.Length, _current.Length - _offset);
            _current.AsSpan(_offset, count).CopyTo(buffer.Span);
            _offset += count;
            return count;
        }

        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
        {
            if (!_segments.Writer.TryWrite(buffer.ToArray()))
            {
                throw new IOException("The seam is closed for writing.");
            }

            return ValueTask.CompletedTask;
        }

        public override Task FlushAsync(CancellationToken cancellationToken) => Task.CompletedTask;

        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
        public override void Write(byte[] buffer, int offset, int count) => throw new NotSupportedException();
        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
        public override void SetLength(long value) => throw new NotSupportedException();
        public override void Flush() { }
    }
}

File: Minimal/tests/Brontide.Minimal.Interchange.Tests/Portable/PortableTestHarness.fs
namespace Brontide.Minimal.Interchange.Tests.Portable

open System
open System.IO
open System.Text.Json
open System.Threading
open System.Threading.Channels
open System.Threading.Tasks
open NUnit.Framework
open Brontide.Minimal.Binding.Portable

/// One direction of a seam: a byte stream whose reader blocks until bytes arrive, observes
/// cancellation, and sees a clean end once the writer completes.
///
/// An anonymous pipe would be closer to the real transport, but its reads neither observe
/// cancellation nor separate "writer closed" from "handle disposed" on Windows. Those are exactly
/// the two distinctions these vectors turn on, so the seam models them explicitly.
type SeamStream() =
    inherit Stream()

    let segments = Channel.CreateUnbounded<byte array>()
    let mutable current: byte array = Array.empty
    let mutable offset = 0

    member _.CompleteWriting() = segments.Writer.TryComplete() |> ignore

    override _.CanRead = true
    override _.CanSeek = false
    override _.CanWrite = true
    override _.Length = raise (NotSupportedException())

    override _.Position
        with get () = raise (NotSupportedException())
        and set (_: int64) = raise (NotSupportedException())

    override _.Flush() = ()
    override _.Read(_: byte array, _: int, _: int) = raise (NotSupportedException())
    override _.Write(_: byte array, _: int, _: int) = raise (NotSupportedException())
    override _.Seek(_: int64, _: SeekOrigin) = raise (NotSupportedException())
    override _.SetLength(_: int64) = raise (NotSupportedException())
    override _.FlushAsync(_: CancellationToken) = Task.CompletedTask

    override _.ReadAsync(buffer: Memory<byte>, token: CancellationToken) =
        ValueTask<int>(
            task {
                let mutable waiting = true
                let mutable ended = false

                while waiting && offset = current.Length do
                    let! available = segments.Reader.WaitToReadAsync token

                    if not available then
                        waiting <- false
                        ended <- true
                    else
                        match segments.Reader.TryRead() with
                        | true, segment ->
                            current <- segment
                            offset <- 0
                            waiting <- false
                        | _ -> ()

                if ended then
                    return 0
                else
                    let count = min buffer.Length (current.Length - offset)
                    ReadOnlyMemory<byte>(current, offset, count).CopyTo buffer
                    offset <- offset + count
                    return count
            }
        )

    override _.WriteAsync(buffer: ReadOnlyMemory<byte>, _: CancellationToken) =
        if segments.Writer.TryWrite(buffer.ToArray()) then
            ValueTask()
        else
            raise (IOException "The seam is closed for writing.")

/// A duplex seam between a host and a provider endpoint inside one test process.
///
/// It is a real pipe pair with real framing, so the process realization's encoding, bounds, and
/// copy accounting are exercised. It is not process isolation: the cross-process evidence lives in
/// the gated test that launches the provider executable.
type PortableLocalSeam(limits: PortableLimits) =
    let hostToProvider = new SeamStream()
    let providerToHost = new SeamStream()
    let cancellation = new CancellationTokenSource()
    let mutable provider: Task = Task.CompletedTask

    member _.HostDuplex: IPortableDuplex =
        PortableStreamDuplex(providerToHost, hostToProvider, limits, false)

    member _.ProviderDuplex: IPortableDuplex =
        PortableStreamDuplex(hostToProvider, providerToHost, limits, false)

    member this.StartProvider(endpoint: PortableProviderEndpoint) =
        provider <- PortableProviderProcessLoop.run this.ProviderDuplex endpoint limits :> Task

    /// Runs a deliberately misbehaving peer instead of the conforming endpoint.
    member this.StartPeer(peer: IPortableDuplex -> CancellationToken -> Task) =
        provider <- peer this.ProviderDuplex cancellation.Token

    /// Ends the provider's outbound direction, which the host observes as a clean end between
    /// frames rather than as an interrupted one.
    member _.CloseProviderOutput() = providerToHost.CompleteWriting()

    member _.CloseProviderInput() = hostToProvider.CompleteWriting()

    interface IDisposable with
        member _.Dispose() =
            cancellation.Cancel()
            hostToProvider.CompleteWriting()
            providerToHost.CompleteWriting()

            try
                provider.Wait(TimeSpan.FromSeconds 5.0) |> ignore
            with _ ->
                // Teardown: a peer failing as its seam disappears is the expected outcome, and the
                // test has already asserted whatever it came to assert.
                ()

            cancellation.Dispose()
            hostToProvider.Dispose()
            providerToHost.Dispose()

[<AutoOpen>]
module PortableTestHarness =

    /// The neutral artifacts are authored data, so a missing string there is a broken fixture
    /// rather than a case the tests should tolerate.
    let str (value: string | null) =
        match value with
        | null -> failwith "The neutral artifact declares a text value that is absent."
        | value -> value

    /// NUnit's multiple-assertion block takes a delegate; naming it keeps the call sites readable.
    let assertAll (assertions: unit -> unit) = Assert.Multiple(Action assertions)

    /// NUnit's equality constraint is ambiguous over F# lists, sets, maps, and arrays, so structural
    /// comparison happens here and the values are reported when it fails.
    let shouldEqual (expected: 'T) (actual: 'T) =
        Assert.That(
            (actual = expected),
            Is.True,
            $"Expected %A{expected}{Environment.NewLine}but was  %A{actual}"
        )

    let neutralRoot =
        Path.Combine(TestContext.CurrentContext.TestDirectory, "binding", "portable")

    let readNeutral (parts: string list) =
        let path = Path.Combine(Array.ofList (neutralRoot :: parts))
        Assert.That(File.Exists path, Is.True, $"The neutral artifact '{path}' was not copied to the test output.")
        JsonDocument.Parse(File.ReadAllText path)

    /// Every vector id declared by the neutral layer, across all vector files.
    let neutralVectorIds () =
        Directory.GetFiles(Path.Combine(neutralRoot, "vectors"), "*.json")
        |> Array.collect (fun file ->
            use document = JsonDocument.Parse(File.ReadAllText file)

            match document.RootElement.TryGetProperty "vectors" with
            | true, vectors -> vectors.EnumerateArray() |> Seq.map (fun vector -> vector.GetProperty("id").GetString() |> str) |> Array.ofSeq
            | _ -> Array.empty)
        |> List.ofArray

    let expectOk (result: PortableResult<'T>) =
        match result with
        | Ok value -> value
        | Error(Refused fault) ->
            Assert.Fail($"Expected success but the layer refused with '{ProtocolCategory.token fault.Category}': {fault.Message}")
            failwith "unreachable"
        | Error(Interrupted failure) ->
            Assert.Fail($"Expected success but the seam reported '{ProcessCategory.token failure.Category}': {failure.Message}")
            failwith "unreachable"

    let expectRefusal (result: PortableResult<'T>) =
        match result with
        | Error(Refused fault) -> fault
        | Error(Interrupted failure) ->
            Assert.Fail($"Expected a refusal but the seam reported '{ProcessCategory.token failure.Category}'.")
            failwith "unreachable"
        | Ok _ ->
            Assert.Fail "Expected a refusal but the layer accepted the input."
            failwith "unreachable"

    /// Asserts the exact portable category a vector states, which is the whole point of the
    /// taxonomy: a refusal is right only when its category is right.
    let expectCategory category (result: PortableResult<'T>) =
        let fault = expectRefusal result
        Assert.That(ProtocolCategory.token fault.Category, Is.EqualTo(ProtocolCategory.token category))
        fault

    let expectProcessFailure (result: PortableResult<'T>) =
        match result with
        | Error(Interrupted failure) -> failure
        | Error(Refused fault) ->
            Assert.Fail($"Expected a process failure but the layer refused with '{ProtocolCategory.token fault.Category}'.")
            failwith "unreachable"
        | Ok _ ->
            Assert.Fail "Expected a process failure but the interaction completed."
            failwith "unreachable"

    let permitted =
        PortableConstraint.AllOf
            [ PortableConstraint.Atom PortableTruth.Satisfied
              PortableConstraint.Atom PortableTruth.Satisfied ]

    let blob (content: string) =
        PortableResource.blob "profile" (Text.Encoding.UTF8.GetBytes content)

    let hex (bytes: byte array) = PortableHex.encode bytes

    let fromHex (value: string) = PortableHex.decode value

    let catalogOf (document: ContractDocument) = expectOk (ShapeCatalog.fromContract document)

    /// The fixed direct-call realization: no wire, no encoding, no copy.
    let directHost (document: ContractDocument) (handler: IPortableOperationHandler) =
        let endpoint = PortableProviderEndpoint(document, handler, Realization.FixedDirectCall)

        PortableBindingHost
            .Establish(document, PortableDirectConversation endpoint, "minimal-direct")
            .Result
        |> expectOk

    let directCoolingHost () =
        directHost CoolingFixture.contract (CoolingHandler())

    /// The negotiated process realization over a local duplex seam: real length-delimited framing
    /// and a real deterministic-CBOR encode and decode on both sides.
    let processHost (document: ContractDocument) (handler: IPortableOperationHandler) =
        let seam = new PortableLocalSeam(document.Limits)
        let endpoint = PortableProviderEndpoint(document, handler, Realization.NegotiatedProcess)
        seam.StartProvider endpoint

        let host =
            PortableBindingHost
                .Establish(document, PortableProcessConversation(seam.HostDuplex, document.Limits), "minimal-process")
                .Result
            |> expectOk

        host, seam

    let processCoolingHost () =
        processHost CoolingFixture.contract (CoolingHandler())

    let invoke (host: PortableBindingHost) operation shape value =
        host.Invoke(operation, shape, value, permitted).Result

    /// Reads a golden encoding's explicit value description into a shaped value. The description
    /// carries kind discriminators; the wire form does not, which is the point of the comparison.
    let rec readDescribedValue (element: JsonElement) : PortableValue =
        match element.GetProperty("kind").GetString() |> str with
        | "unit" -> PortableUnit
        | "text" -> PortableText(element.GetProperty("value").GetString() |> str)
        | "boolean" -> PortableBoolean(element.GetProperty("value").GetBoolean())
        | "integer" -> PortableInteger(element.GetProperty("value").GetInt64())
        | "decimal" ->
            PortableDecimalValue(element.GetProperty("exponent").GetInt32(), element.GetProperty("mantissa").GetInt64())
        | "bytes" -> PortableBytesValue(fromHex (element.GetProperty("hex").GetString() |> str))
        | "sequence" ->
            PortableSequence(element.GetProperty("items").EnumerateArray() |> Seq.map readDescribedValue |> List.ofSeq)
        | "choice" ->
            PortableChoice(
                element.GetProperty("case").GetString() |> str,
                readDescribedValue (element.GetProperty "value")
            )
        | "record" ->
            let fields =
                element.GetProperty("fields").EnumerateObject()
                |> Seq.map (fun field -> field.Name, readDescribedValue field.Value)
                |> Map.ofSeq

            let fragments =
                element.GetProperty("fragments").EnumerateObject()
                |> Seq.map (fun fragment ->
                    let reference = expectOk (PortableFragmentRef.tryParseText fragment.Name)

                    let fields =
                        fragment.Value.EnumerateObject()
                        |> Seq.map (fun field -> field.Name, readDescribedValue field.Value)
                        |> Map.ofSeq

                    reference, fields)
                |> Map.ofSeq

            PortableRecord(fields, fragments)
        | other -> failwith $"The golden value description uses unknown kind '{other}'."

    let readDescribedShape (element: JsonElement) =
        expectOk (PortableShapeRef.tryCreate (element.GetProperty("name").GetString() |> str) (element.GetProperty("version").GetInt32()))
